// @roots/sage/lib/cjs/theme/plugin.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThemeJsonWebpackPlugin = void 0;
const bud_support_1 = require("@roots/bud-support");
/**
 * ThemeJSONWebpackPlugin
 *
 * @public
 */
class ThemeJsonWebpackPlugin {
    /**
     * Class constructor
     *
     * @param options - Plugin options
     *
     * @public
     */
    constructor(options) {
        this.options = options;
    }
    /**
     * theme.json path
     *
     * @public
     */
    get path() {
        return this.options.path;
    }
    /**
     * theme.json settings
     *
     * @public
     */
    get settings() {
        return JSON.stringify({
            __generated__: '⚠️ This file is generated. Do not edit.',
            $schema: 'https://schemas.wp.org/trunk/theme.json',
            version: 2,
            settings: this.options.settings,
        }, null, 2);
    }
    /**
     * Apply plugin
     *
     * @param compiler - Webpack compiler
     * @returns void
     *
     * @public
     * @decorator `@bind`
     */
    apply(compiler) {
        compiler.hooks.done.tapPromise(this.constructor.name, this.done);
    }
    /**
     * Compiler done
     *
     * @returns Promise
     *
     * @public
     * @decorator `@bind`
     */
    async done() {
        try {
            await bud_support_1.fs.writeFile(this.path, this.settings, 'utf8');
        }
        catch (err) {
            throw new Error(err);
        }
    }
}
__decorate([
    bud_support_1.bind
], ThemeJsonWebpackPlugin.prototype, "apply", null);
__decorate([
    bud_support_1.bind
], ThemeJsonWebpackPlugin.prototype, "done", null);
exports.ThemeJsonWebpackPlugin = ThemeJsonWebpackPlugin;
