// @roots/bud-server/lib/cjs/server/server.https.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Https = void 0;
const bud_support_1 = require("@roots/bud-support");
const https_1 = require("https");
const server_base_1 = require("./server.base");
const { readFile } = bud_support_1.fs;
const { isUndefined } = bud_support_1.lodash;
/**
 * HTTPS Server
 * @public
 */
class Https extends server_base_1.BaseServer {
    /**
     * Has options
     * @returns boolean
     * @public
     * @decorator `@bind`
     */
    hasOptions() {
        return Object.keys(this.options).length > 0;
    }
    /**
     * Has SSL key
     * @public
     */
    hasKey() {
        return this.hasOptions() && !isUndefined(this.options.key);
    }
    /**
     * Has SSL certificate
     * @public
     */
    hasCert() {
        return this.hasOptions() && !isUndefined(this.options.cert);
    }
    /**
     * Get SSL key
     * @returns
     */
    async getKey() {
        if (!this.hasKey()) {
            this.app.warn('Server key is not defined');
            return;
        }
        try {
            return await readFile(this.options.key, 'utf8');
        }
        catch (err) {
            this.app.error(err);
        }
    }
    /**
     * Get SSL certificate
     * @public
     * @decorator `@bind`
     */
    async getCert() {
        if (!this.hasKey()) {
            this.app.warn('Server key is not defined');
            return;
        }
        try {
            return await readFile(this.options.cert, 'utf8');
        }
        catch (err) {
            this.app.error(err);
        }
    }
    /**
     * Create HTTPS server
     * @public
     * @decorator `@bind`
     */
    async createServer(express) {
        if (!this.hasOptions()) {
            this.instance = (0, https_1.createServer)(express);
            return this.instance;
        }
        const key = await this.getKey();
        const cert = await this.getCert();
        const options = { ...this.options, ...({ key } ?? {}), ...({ cert } ?? {}) };
        this.instance = (0, https_1.createServer)(options, express);
        return this.instance;
    }
}
__decorate([
    bud_support_1.bind
], Https.prototype, "hasOptions", null);
__decorate([
    bud_support_1.bind
], Https.prototype, "hasKey", null);
__decorate([
    bud_support_1.bind
], Https.prototype, "hasCert", null);
__decorate([
    bud_support_1.bind
], Https.prototype, "getKey", null);
__decorate([
    bud_support_1.bind
], Https.prototype, "getCert", null);
__decorate([
    bud_support_1.bind
], Https.prototype, "createServer", null);
exports.Https = Https;
